File: examples/ts/pick.ts
import crypto from 'crypto';

async function callPickApi(path: string, body: object) {
    try {
        if (!process.env.PICK_API_URL || !process.env.PICK_API_KEY) {
            throw new Error("PICK_API_URL or PICK_API_KEY is not set");
        }
        const response = await fetch(`${process.env.PICK_API_URL}${path}`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${process.env.PICK_API_KEY}`,
            },
            body: JSON.stringify(body),
        });

        if (response.status >= 400) {
            throw new Error(`HTTP ${response.status} ${response.statusText}`);
        }
        
        return response.json();
    } catch (e: any) {
        console.log(`[pick.biq.blue] Error ${e.message ? e.message : 'unknown'} at ${path}`);
    }
    return null;
}

async function shouldPick(hash: string): Promise<string> {
    const res = await callPickApi("/pick", { hash });
    return res && res.pick === "ON-DEMAND" ? "ON-DEMAND" : "RESERVATION";
}

export async function getFromHash(hash: string): Promise<string> {
    return shouldPick(hash);
}

export async function getFromQuery(query: string): Promise<string> {
    const hash = crypto.createHash("sha256").update(query).digest("hex");
    return getFromHash(hash);
}

export async function updateFromHash(hash: string, job: any) {
    if (!job.id) {
        console.log(`[pick.biq.blue] No job id for ${hash}`);
    } else {
        await callPickApi("/write", {
            hash,
            job_id: job.id,
            creation_time: job.metadata.statistics.creationTime,
            start_time: job.metadata.statistics.startTime,
            end_time: job.metadata.statistics.endTime,
            total_slot_ms: job.metadata.statistics.totalSlotMs,
            total_bytes_billed: job.metadata.statistics.query.totalBytesBilled,
            total_bytes_processed: job.metadata.statistics.query.totalBytesProcessed,
            bi_engine_mode: job.metadata.statistics.query.biEngineStatistics?.biEngineMode || '',
            reservation_id: null,
        });
    }
}

export async function updateFromQuery(query: string, job: any) {
    const hash = crypto.createHash("sha256").update(query).digest("hex");
    return updateFromHash(hash, job);
}

File: examples/python/pick.py
import hashlib
import os
import requests

def call_pick_api(path, body):
    try:
        api_url = os.getenv("PICK_API_URL")
        api_key = os.getenv("PICK_API_KEY")
        if not api_url or not api_key:
            raise ValueError("PICK_API_URL or PICK_API_KEY is not set")

        response = requests.post(
            f"{api_url}{path}",
            json=body,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}",
            }
        )
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"[pick.biq.blue] Error {e} at {path}")
        return None

async def should_pick(hash):
    res = call_pick_api("/pick", {"hash": hash})
    if not res or not res.get("pick"):
        return "ON-DEMAND"  # Défaut à "ON-DEMAND"
    return res["pick"]

async def write_query(body):
    call_pick_api("/write", body)

# GET
async def get_from_hash(hash):
    return await should_pick(hash)

async def get_from_query(query):
    # Calculer le hash de la requête
    hash = hashlib.sha256(query.encode()).hexdigest()
    return await get_from_hash(hash)

# UPDATE
async def update_from_hash(hash, job):
    if not job.job_id:
        print(f"[pick.biq.blue] No job id for {hash}")
    else:
        await write_query({
            "hash": hash,
            "job_id": job.job_id,
            "creation_time": job.created.isoformat(),
            "start_time": job.started.isoformat(),
            "end_time": job.ended.isoformat(),
            "total_slot_ms": job.total_slot_ms,
            "total_bytes_billed": job.total_bytes_billed,
            "total_bytes_processed": job.total_bytes_processed,
            "bi_engine_mode": getattr(job, "bi_engine_statistics", {}).get("bi_engine_mode", None),
            "reservation_id": None,
        })

async def update_from_query(query, job):
    hash = hashlib.sha256(query.encode()).hexdigest()
    await update_from_hash(hash, job)
